Web Inspector: adopt more smart pointers in InspectorAnimationAgent (part 2)

burg opened the pull request:
#### bb6619c2421179b0329fd2a40e51af9b24657a88
<pre>
Web Inspector: adopt more smart pointers in InspectorAnimationAgent (part 2)
<a href="https://bugs.webkit.org/show_bug.cgi?id=300926">https://bugs.webkit.org/show_bug.cgi?id=300926</a>

Reviewed by Devin Rousso.

No new tests, no behavior change.

* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::InspectorAnimationAgent::resolveAnimation):
Do not resolve/bind an animation if its ScriptExecutionContext is gone.
(WebCore::InspectorAnimationAgent::findAnimationId):
(WebCore::InspectorAnimationAgent::bindAnimation):

* Source/WebCore/inspector/agents/InspectorAnimationAgent.h:
Use a WeakRef since we don&apos;t expect these references to become null.
We nonetheless should not extend the lifetime of the WebAnimation objects.

Canonical link: <a href="https://commits.webkit.org/302497@main">https://commits.webkit.org/302497@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1636b4dcc3669d5249a4469da5792673089eca12

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129267 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1523 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40103 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136642 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80659 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/848bc257-42ca-411e-ab0e-b8318f981628) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131138 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1456 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1400 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98436 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66338 "") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/620cddc4-edd0-4d7d-ab27-8c1ea4acef4d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132214 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1152 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115795 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79087 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/49d11798-e427-4ea3-9711-af6fdc39a6f1) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1063 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33914 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79921 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109516 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34409 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139116 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1314 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1264 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106965 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1363 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112131 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106803 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27195 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1085 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30650 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53937 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1386 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64742 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1209 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1245 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1308 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/agents/InspectorAnimationAgent.h:
```diff
@@ -100,7 +100,7 @@ class InspectorAnimationAgent final : public InspectorAgentBase, public Inspecto
     WeakRef<Page> m_inspectedPage;
 
-    MemoryCompactRobinHoodHashMap<Inspector::Protocol::Animation::AnimationId, WebAnimation*> m_animationIdMap;
+    MemoryCompactRobinHoodHashMap<Inspector::Protocol::Animation::AnimationId, Ref<WebAnimation>> m_animationIdMap;
```
we absolutely do not want to do this

the frontend does not have any mechanism for telling the backend that it no longer wants to instrument a `WebAnimation`

in fact, the frontend relies on `Animation.animationDestroyed` events from the backend (specifically in `WebAnimation::~WebAnimation`) to know when things have been destroyed (i.e. the frontend has full knowledge of every `WebAnimation` from just after it's created to right before it's destroyed)

as a more general point, Web Inspector instrumentation should rarely if ever keep anything alive as that would change the behavior of the page (vs when Web Inspector is not open), which goes against the ethos of Web Inspector being as non-impactful as possible (both performance and effect)

if we have to make some sort of change here then it should really be a `WeakRef` (though id echo the same "concern" as <https://github.com/WebKit/WebKit/pull/51516#pullrequestreview-3297326007> since it isn't possible for this map to contain anything that isn't alive since the instrumentation points are in the constructor and destructor of `WebAnimation`)

dcrousso commented on Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
```diff
@@ -362,6 +362,9 @@ Inspector::Protocol::ErrorStringOr<Ref<Inspector::Protocol::Runtime::RemoteObjec
+    if (!animation->scriptExecutionContext())
+        return makeUnexpected("Context for Animation is gone"_s);
+
     auto* state = animation->protectedScriptExecutionContext()->globalObject();
```
is this actually necessary?  id expect the `WebAnimation` to be destroyed if the `ScriptExecutionContext` doesn't exist anymore

if it is, i feel like we should do something like this instead
```cpp
RefPtr scriptExecutionContext = animation->scriptExecutionContext();
if (!scriptExecutionContext)
    return makeUnexpected("Animation for given animationId is detached"_s);

auto* state = scriptExecutionContext->globalObject();
```

burg committed:
Web Inspector: adopt more smart pointers in InspectorAnimationAgent (part 2)
https://bugs.webkit.org/show_bug.cgi?id=300926

Reviewed by Devin Rousso.

No new tests, no behavior change.

* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
(WebCore::InspectorAnimationAgent::resolveAnimation):
Do not resolve/bind an animation if its ScriptExecutionContext is gone.
(WebCore::InspectorAnimationAgent::findAnimationId):
(WebCore::InspectorAnimationAgent::bindAnimation):

* Source/WebCore/inspector/agents/InspectorAnimationAgent.h:
Use a WeakRef since we don't expect these references to become null.
We nonetheless should not extend the lifetime of the WebAnimation objects.

Canonical link: https://commits.webkit.org/302497@main
